### PR TITLE
Enhanced S3 APIs for ECS extension features

### PR DESCRIPTION
**Pull request by yangb8**

### CODE REVIEW

File: api.go
Comment (DavidASeibert) [inline api.go:1]: You should have copyrights in all files.

File: api.go
Comment (DavidASeibert) [inline api.go]: Probably should spell out all of Ext - Extension? Exterior? Extrusion? ;-)

This should apply everywhere you use Ext.

Or, you could use Ecs/ECS to make it even clearer that this is the ECS-specific version. That's probably even better.

File: api.go
Comment (yangb8) [reply api.go:1]: accepted

File: api.go
Comment (DavidASeibert) [inline api.go]: There are a number of metadata misspellings as "Metdata". Please correct those.

File: api.go
Comment (DavidASeibert) [inline api.go]: Why "Extg"? Should that be "Ext"? It's not clear to me what this means, so use a better name if it's not.

File: api.go
Comment (DavidASeibert) [inline api.go:402]: I'd move this closer to where it's used in the CreateBucket methods.

File: api.go
Comment (DavidASeibert) [inline api.go]: Is the Emc necessary? If so, we should use EmcEcs everywhere we have our own variant, for consistency. If not, we should use Ecs for all.

File: api.go
Comment (DavidASeibert) [inline api.go]: I'd drop the Ext after EmcEcs (or just Ecs) everywhere, it's redundant (or should be).

**Review by DavidASeibert — CHANGES_REQUESTED** [review]
One high level pass done, this file only

File: api.go
Comment (yangb8) [reply api.go]: accepted. will change it to "Extension". unnecessary to add ecs keyword because the github URL includes emcecs.

File: api.go
Comment (yangb8) [reply api.go]: accepted.

File: api.go
Comment (yangb8) [reply api.go]: accepted. will delete it. this function "HeadBucketExtg" is junk. "HeadBucketExt" is valid one and in this file as well.

File: api.go
Comment (yangb8) [reply api.go:402]: accepted

File: api.go
Comment (yangb8) [reply api.go]: accepted

File: api.go
Comment (yangb8) [reply api.go]: accepted

File: api.go
Comment (yangb8) [reply api.go:402]: I looked into further, actually,  xxxInput and xxxOutput structures are located after xxxRequest functions. I'd prefer to keep it where it is. (the location sequence in api.go is all request functions --> input/output structures --> misc util functions)

**Comment by yangb8** [comment]
@DavidASeibert any further comments?

File: api.go
Comment (DavidASeibert) [inline api.go]: "Mds" should be Metadata, it's spelled out in full everywhere else.

File: api.go
Comment (DavidASeibert) [inline api.go]: As above - "Mds" and "Md" should be Metadata, that's spelled out in full everywhere else. There are more instances below, please find and fix them all. Once that's done, it's approved.

**Review by DavidASeibert — APPROVED** [review]
There are still a bunch of places where "Md" is used, apparently for "Metadata". Please change those to use the full word, so developers don't need to even think about what the abbreviation might be. It's approved after that.

File: api.go
Comment (yangb8) [reply api.go]: `QueryMds` is the field name in response body of metadata search, that’s why I used that name. So I’ll leave it as it is if you are ok with that.

File: api.go
Comment (yangb8) [reply api.go]: `QueryMds` is the field name in response body of metadata search, that’s why I used that name. So I’ll leave it as it is if you are ok with that.

File: api.go
Comment (yangb8) [reply api.go]: after discussion with David on slack. the change will be "EcsQueryMds -> QueryMetadata (or EcsQueryMetada) and MdType -> MetadataType (or EcsMetadataType), and leave the other two."

**Pull request merged**
